File: cpp/testsuite/OfflineExamples/RaytracingOfflineExample/src/ImageExporter.cpp
#include "ImageExporter.h"

#include "vsdk/toolkit/io/image/ImagePersistence.h"
#include "java/io/File.h"
#include "vsdk/toolkit/media/RGBImageUncompressed.h"

#include <cctype>
#include <cstdio>
#include <cstring>

bool ImageExporter::exportImage(const java::String& outputFileName, RGBImageUncompressed* image)
{
    java::File outFile(outputFileName.c_str());
    printf("Exporting result image to file \"%s\": ", outputFileName.c_str());

    java::String lower = outputFileName;
    for (size_t i = 0; i < lower.size(); i++) {
        lower[i] = std::tolower((unsigned char)lower[i]);
    }

    bool ok = false;
    if ( lower.size() >= 4 && lower.substr(lower.size()-4) == ".png" ) {
        ok = ImagePersistence::exportPNG(outFile, image);
    }
    else if ( (lower.size() >= 4 && lower.substr(lower.size()-4) == ".jpg") ||
              (lower.size() >= 5 && lower.substr(lower.size()-5) == ".jpeg") ) {
        ok = ImagePersistence::exportJPEG(outFile, image);
    }
    else {
        ok = ImagePersistence::exportPPM(outFile, image);
    }

    if ( ok ) {
        printf(" OK!\n");
    }
    return ok;
}

File: typescript/testsuite/OfflineExamples/RaytracingOfflineExample/src/ImageExporter.ts
import { platformPrint, platformPrintln } from "@vitral/base";
import type { RGBImageUncompressed } from "@vitral/base";
import { File, ImagePersistence } from "@vitral/fs";

export class ImageExporter {
    public export(outputFileName: string, image: RGBImageUncompressed): boolean {
        const outputFile: File = new File(outputFileName);

        platformPrint('Exporting result image to file "' + outputFileName + '": ');
        const exported: boolean = this.exportToFile(outputFile, image);
        if (exported) {
            platformPrintln(" OK!");
        }
        return exported;
    }

    /**
    Java overloads `export` on `String` / `File`; TypeScript overload
    resolution on two one-argument-different signatures would be ambiguous at
    the call site, so the private half keeps a distinct name.
    */
    private exportToFile(outputFile: File, image: RGBImageUncompressed): boolean {
        // Java uses `toLowerCase(Locale.ROOT)`; JavaScript's argument-less
        // `toLowerCase` is already the locale-independent mapping.
        const lowerName: string = outputFile.getName().toString().toLowerCase();
        if (lowerName.endsWith(".png")) {
            ImagePersistence.exportPNG(outputFile, image);
            return true;
        }
        if (lowerName.endsWith(".jpg") || lowerName.endsWith(".jpeg")) {
            // Java delegates to `ImagePersistence.exportJPG`, which is backed
            // by javax.imageio. No JPEG encoder is ported to TypeScript yet,
            // so this branch reports the missing capability rather than
            // silently writing a PPM under a .jpg name.
            throw new Error("JPEG export is not ported to TypeScript yet; use a .png or .ppm output file name.");
        }
        return ImagePersistence.exportPPM(outputFile, image);
    }
}
